libs/langchain/langchain/agents/agent.py

```diff
@@ -1263,3 +1266,74 @@ def _prepare_intermediate_steps(
             return self.trim_intermediate_steps(intermediate_steps)
         else:
             return intermediate_steps
+
+    def stream(
+        self,
+        input: Dict[str, Any],
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> Iterator[AgentStreamOutput]:
+        """Enables streaming over steps taken to reach final output."""
+        config = config or {}
+        iterator = AgentExecutorIterator(
+            self,
+            input,
+            config.get("callbacks"),
+            tags=config.get("tags"),
+            metadata=config.get("metadata"),
+            run_name=config.get("run_name"),
+            **kwargs,
+        )
+        for step in iterator:
+            if "intermediate_step" in step:
+                yield AgentStreamOutput(
+                    actions=[a for a, o in step["intermediate_step"]],
+                    observations=[o for a, o in step["intermediate_step"]],
+                )
+            else:
+                yield AgentStreamOutput(step)
+
+    async def astream(
+        self,
+        input: Dict[str, Any],
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> AsyncIterator[AgentStreamOutput]:
+        """Enables streaming over steps taken to reach final output."""
+        config = config or {}
+        iterator = AgentExecutorIterator(
+            self,
+            input,
+            config.get("callbacks"),
+            tags=config.get("tags"),
+            metadata=config.get("metadata"),
+            run_name=config.get("run_name"),
+            **kwargs,
+        )
+        async for step in iterator:
+            if "intermediate_step" in step:
+                yield AgentStreamOutput(
+                    actions=[a for a, o in step["intermediate_step"]],
+                    observations=[o for a, o in step["intermediate_step"]],
+                )
+            else:
+                yield AgentStreamOutput(step)
+
+
+class AgentStreamOutput(AddableDict):
+    """dict subclass for agent stream output.
+
+    Implements addition to allow for easy concatenation of stream outputs.
+
+    Additionally, adds a `intermediate_steps` key that contains the
+    combination of `actions` and `observations` from the stream output.
+    """
+
+    def __missing__(self, key: str) -> Any:
+        if key == "intermediate_steps":
```

Comment:
this is plural but in stream it's singular

Reply:
nvm think i get it (though it is confusing)

Reply:
It's the name in agent iterator that is odd imo, but I can't change that one (breaking change)

Reply:
wow didn't know about __missing__